src/compiler/moduleSpecifiers.ts

```diff
@@ -453,6 +457,23 @@ namespace ts.moduleSpecifiers {
                 else if (pattern === relativeToBaseUrl || pattern === relativeToBaseUrlWithIndex) {
                     return key;
                 }
+                else if (resolvedPaths) {
+                    const resolvedCandidate = resolvedPaths?.[key]?.[i];
+                    if (resolvedCandidate) {
+                        if (redirectFileName === resolvedCandidate) {
+                            return key;
+                        }
+                        const relativeFileName = removeFileExtension(normalizePath(getRelativePathToDirectoryOrUrl(

```

Comment:
Why is this different from when we find the * in the pattern. It should be something similar string manipulation as there is only little difference when star is found or not.
https://github.com/microsoft/TypeScript/blob/master/src/compiler/moduleNameResolver.ts#L1373

Reply:
If there’s a `*` then we can’t resolve this ahead of time since we don’t know the path with the substitution. I’m not going to do all that work during module specifier resolution because that happens during completions. I guess the only good way to do this would be to keep a cache mapping package root directories to their `types`/`main` file paths. But I don’t think this feature/bugfix merits any more complexity than this, so I would rather close it as wontfix and tell people not to point paths at package roots if they want auto imports to work correctly.